File: code/track_objects.m
function track_objects()

% this is a very simple tracking algo
% for more serious tracking, look-up the papers in the projects pdf
FRAME_DIR = '../data/frames/';
DET_DIR = '../data/detections/';
start_frame = 62;
end_frame = 71;
%tracks = zeros(0,13);   %formart [dets_cur dets_next occurences]

%field names
previous_frame = 'prev_frame';  previous_frame_val = zeros(4,0);  %had to transpose so ismember would work
occurences = 'occur';  occurences_val = {0};
dets = 'dets';  dets_val = zeros(0,6);
frames_field = 'frames'; frames_val = zeros(0,0);  % the frames a detection occurs in 
tracks = struct(previous_frame,previous_frame_val,occurences,occurences_val,dets,dets_val,frames_field,frames_val);


    for idx = start_frame:end_frame
        im_cur = imread(fullfile(FRAME_DIR, sprintf('%06d.jpg', idx)));
        data = load(fullfile(DET_DIR, sprintf('%06d_dets.mat', idx)));
        dets_cur = data.dets;

        im_next = imread(fullfile(FRAME_DIR, sprintf('%06d.jpg', idx+1)));
        data = load(fullfile(DET_DIR, sprintf('%06d_dets.mat', idx+1)));
        dets_next = data.dets;

        % sim has as many rows as dets_cur and as many columns as dets_next
        % sim(k,t) is similarity between detection k in frame i, and detection
        % t in frame j
        % sim(k,t)=0 means that k and t should probably not be the same track
        sim = compute_similarity(dets_cur, dets_next, im_cur, im_next);

        %greedy approach 
        finding = true;   %loop until no values in sim are > 0
        while finding
            [maxValue, idxs] = max(sim(:));  %find current max
            [i,j] = ind2sub(size(sim),idxs);  % get its i and j

            if maxValue == 0 
                finding = false; %stop loop all sims>0 are found
            else
                %get previous dets locations from tracks
                prev_dets = [tracks.prev_frame]'; % transpose so that ismember function could find row

                %see if current det is in tracks by comparing with prev_dets
                trans_dets_cur = dets_cur(i,1:4); % get current sims detection  box
                [Result,LocResult] = ismember(trans_dets_cur,prev_dets,'rows');

                if Result  %if current det is in tracks, then add to that track
                    curr_occurs = tracks(LocResult).occur + 1; %increase number of occurence for current track by 1
                    curr_dets = tracks(LocResult).dets;    
                    previous_frame_val = dets_next(j,1:4)';   %set prev_fram value to det_next 
                    dets_val = [curr_dets;dets_next(j,1:4)];  %append the next det to the tracks dets list
                    frames_val = [tracks(LocResult).frames idx+1];  %save the frames this track was seen in 
                    
                    tracks(LocResult) = struct(previous_frame,previous_frame_val,occurences,curr_occurs,dets,dets_val,frames_field,frames_val);
                    %tracks(LocResult,:) = [dets_cur(i,:) dets_next(j,:) curr_occurs+1];   %save the detections, occured twice (current & next)                             
                    
                else
                    %add current det and next det to a new track
                    track_size = size(tracks,2);      
                    curr_occurs = 2;  
                    previous_frame_val = dets_next(j,1:4)';
                    dets_val = [dets_cur(i,1:4);dets_next(j,1:4)];
                    frames_val = [idx idx+1];
                    
                    tracks(track_size+1) = struct(previous_frame,previous_frame_val,occurences,curr_occurs,dets,dets_val,frames_field,frames_val);
                    %tracks(track_size+1) = [dets_cur(i,:) dets_next(j,:) 2];   %save the detections, occured twice (current & next)         
                end
                sim(i,j)=0; %set the current sim to 0 so it wont be choosen again

            end
        end   

    end;

    
    %only take tracks with occurences > 2 
    idx = find([tracks.occur]>2);
    correct_tracks = tracks(idx);
    
    %visualize tracks with more than 5
    idx = find([tracks.occur]>5);
    visualize_tracks = tracks(idx);
    track_size = size(visualize_tracks,2);
    colors = {'r','g','b', 'y', 'm', 'c', 'w','k'}; %set colors for modified showboxes function
    
    %%loop through each frame and show detections 
    for idx = start_frame:end_frame
        im_cur = imread(fullfile(FRAME_DIR, sprintf('%06d.jpg', idx)));
        boxes = [];
        frame_colors = {};
        
        %get the boxes (loop through each track)
        for i = 1:track_size
            track_dets = visualize_tracks(i).dets; %get dets for all frams on current track
            track_frames = visualize_tracks(i).frames;
            location = find(track_frames == idx);
            
            if location  %if current track has current frame
                boxes = [boxes track_dets(location,:)];
                frame_colors{size(frame_colors,2)+1} = colors{i}; %set a color got current box
            end
        end
        
        figure,showboxes(im_cur, boxes, frame_colors) %run modified showbox function

    end
    
    
end


function sim = compute_similarity(dets_cur, dets_next, im_cur, im_next)

n = size(dets_cur, 1);
m = size(dets_next, 1);
sim = zeros(n, m);


area_cur = compute_area(dets_cur);
area_next = compute_area(dets_next);
c_cur = compute_center(dets_cur);
c_next = compute_center(dets_next);
im_cur = double(im_cur);
im_next = double(im_next);
weights = [1,1,2];

for i = 1: n
    % compare sizes of boxes
    a = area_cur(i) * ones(m, 1);
    sim(i, :) = sim(i, :) + weights(1) * (min(area_next, a) ./ max(area_next, a))';
    
    % penalize distance (would be good to look-up flow, but it's slow to
    % compute for images of this size)
    sim(i, :) = sim(i, :) + weights(2) * exp((-0.5*sum((repmat(c_cur(i, :), [size(c_next, 1), 1]) - c_next).^2, 2)) / 5^2)';
    
    % compute similarity of patches
    box = round(dets_cur(i, 1:4));
    box(1:2) = max([1,1],box(1:2));
    box(3:4) = [min(box(3),size(im_cur, 2)), min(box(4),size(im_cur, 1))];
    im_i = im_cur(box(2):box(4),box(1):box(3), :);
    im_i = im_i / norm(im_i(:));
    for j = 1 : m
       d = norm(c_cur(i, :) - c_next(j, :));
       if d>60  % distance between boxes too big
           sim(i,j) = 0;
           continue;
       end;
       box = round(dets_next(j, 1:4));
       box(1:2) = max([1,1],box(1:2));
       box(3:4) = [min(box(3),size(im_cur, 2)), min(box(4),size(im_cur, 1))]; 
       im_j = im_next(box(2):box(4),box(1):box(3), :);
       im_j = double(imresize(uint8(im_j), [size(im_i, 1), size(im_i, 2)]));
       im_j = im_j / norm(im_j(:));
       c = sum(im_i(:) .* im_j(:));
       sim(i,j) = sim(i,j) + weights(3) * c;
    end;
end;
end

function area = compute_area(dets)
   area = (dets(:, 3) - dets(:, 1) + 1).* (dets(:, 4) - dets(:, 2) + 1);
end

function c = compute_center(dets)

c = 0.5 * (dets(:, [1:2]) + dets(:, [3:4]));
end

File: code/track_objects.py
"""A very simple tracking algorithm.

For more serious tracking, look-up the papers in the projects pdf.
"""

import numpy as np
import scipy.io
import scipy.misc
from PIL import Image
import os

FRAME_DIR = '../data/frames/'
DETECTION_DIR = '../data/detections/'


def compute_similarity(current_detections, next_detections,
                       current_image, next_image):
  """Compute similarity score for detections in adjacent frames.

  Args:
    current_detections: Detections in the current frame
    next_detections:    Detections in the next frame
    current_image:      PIL.Image object for the current frame
    next_image:         PIL.Image object for the next frame

  Returns:
    sim:                Similarity matrix.
  """
  num_current_detections = len(current_detections)
  num_next_detections = len(next_detections)
  sim = np.zeros((num_current_detections, num_next_detections))

  current_detection_areas = compute_area(current_detections)
  next_detection_areas = compute_area(next_detections)

  current_detection_centers = compute_center(current_detections)
  next_detection_centers = compute_center(next_detections)

  current_image = np.array(current_image)
  next_image = np.array(next_image)

  weights = [1, 1, 2]

  for detection_id in range(num_current_detections):
    # Compare boxes sizes
    current_box_area = current_detection_areas[detection_id]
    sim[detection_id] = (np.minimum(next_detection_areas, current_box_area)
                         / np.maximum(next_detection_areas, current_box_area)
                         * weights[0])

    # Penalize distance (would be good to look-up flow, but it's slow to
    # compute for images of this size.
    center_distances = (next_detection_centers
                        - current_detection_centers[detection_id])
    sim[detection_id] += np.exp(- 0.5 * np.sum(center_distances ** 2, axis=1)
                                / 5 ** 2) * weights[1]

    # Compute patch similarity
    current_patch = grab_image_patch(current_image,
                                     current_detections[detection_id])

    for next_detection_id in range(num_next_detections):
      distance = np.linalg.norm(current_detection_centers[detection_id]
                                - next_detection_centers[next_detection_id])
      if distance > 60:
        sim[detection_id, next_detection_id] = 0.
        continue

      next_patch = grab_image_patch(next_image,
                                    next_detections[next_detection_id],
                                    current_patch.shape[:2])

      sim[detection_id, next_detection_id] += (
          weights[2] * np.sum(current_patch * next_patch))
  return sim


def grab_image_patch(image, box, target_shape=None):
  """Get the image patch based on the given box.

  Args:
    image:          np.array object for input image.
    box:            Pixel coordinate for output patch.
    target_shape:   None or size-2 tuple for target patch shape. Output patch
                    will be resized to target shape if not set to None.

  Returns:
    patch:    np.array object for cropped image patch
  """
  box = np.round(box).astype(np.int32)

  # Python index starts at 0.
  box[:2] = np.maximum(1, box[:2]) - 1
  # xmax and ymax are exclusive.
  box[2] = min(box[2], image.shape[1])
  box[3] = min(box[3], image.shape[0])

  patch = image[box[1]:box[3], box[0]:box[2], :]

  if target_shape is not None:
    patch = scipy.misc.imresize(patch, target_shape)

  patch = patch.astype(np.double)
  patch /= np.linalg.norm(patch)

  return patch


def compute_area(detections):
  """Compute the area for each detection.

  Args:
    detections:   A matrix of shape N * 4, recording the pixel coordinate of
                  each detection bounding box (inclusive).

  Returns:
    area:         A vector of length N, representing the area for each input
                  bounding box.
  """
  area = (np.maximum(0, detections[:, 2] - detections[:, 0] + 1)
          * np.maximum(0, detections[:, 3] - detections[:, 1] + 1))
  return area


def compute_center(detections):
  """Compute the center for each detection.

  Args:
    detections:   A matrix of shape N * 4, recording the pixel coordinate of
                  each detection bounding box (inclusive).

  Returns:
    center:       A matrix of shape N I 2, representing the (x, y) coordinate
                  for each input bounding box center.
  """
  center = detections[:, [0, 1]] + detections[:, [2, 3]]
  center /= 2.
  return center


def load_image_and_detections(frame_id):
  """Helper function for loading image and detections.
  """
  image_path = os.path.join(FRAME_DIR, '%06d.jpg' % frame_id)
  image = Image.open(image_path)

  mat_path = os.path.join(DETECTION_DIR, '%06d_dets.mat' % frame_id)
  mat_content = scipy.io.loadmat(mat_path)
  detections = mat_content['dets']
  return image, detections


start_frame = 62
end_frame = 72

sims = []
for frame_id in range(start_frame, end_frame):
  current_image, current_detections = load_image_and_detections(frame_id)

  next_image, next_detections = load_image_and_detections(frame_id + 1)

  # sim has as many rows as len(current_detections) and as many columns as
  # len(next_detections).
  # sim[k, t] is the similarity between detection k in frame i, and detection
  # t in frame j.
  # sim[k, t] == 0 indicates that k and t should probably not be the same track.
  sim = compute_similarity(current_detections, next_detections,
                           current_image, next_image)







